### PR TITLE
promote PodReadinessGate feature to GA

### PR DESCRIPTION
**Pull request by freehan**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR contains code changes necessary to graduate the PodReadinessGate feature from `Beta` to `GA`
Enhancement: https://github.com/kubernetes/enhancements/issues/580
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md

**Does this PR introduce a user-facing change?**:

NONE

```release-note
PodReadinessGate feature is now GA. The feature gate will not allow disabling it.
```

### CODE REVIEW

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -470,7 +470,7 @@ var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureS
 	RunAsGroup:                                  {Default: false, PreRelease: utilfeature.Alpha},
 	VolumeSubpath:                               {Default: true, PreRelease: utilfeature.GA},
 	BalanceAttachedNodeVolumes:                  {Default: false, PreRelease: utilfeature.Alpha},
-	PodReadinessGates:                           {Default: true, PreRelease: utilfeature.Beta},
+	PodReadinessGates:                           {Default: true, PreRelease: utilfeature.GA},
```
Comment (liggitt) [inline pkg/features/kube_features.go]: typically, when promoting a feature gate to GA, we should:
* set `LockToDefault: true`
* add a comment to delete in 2 releases (per https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior, rule 8)
* add a release note that the feature is now enabled, the feature gate will not allow disabling it, and the feature gate will be removed in 1.16

Comment (freehan) [reply pkg/features/kube_features.go]: Thanks! I have made the corresponding changes.

**Comment by liggitt** [comment]
if the gate is locked to enabled, we should drop these:

https://github.com/kubernetes/kubernetes/blob/1eb2acca99fefad25a29870ffc9de29213db943b/pkg/api/pod/util.go#L321-L323

https://github.com/kubernetes/kubernetes/blob/1eb2acca99fefad25a29870ffc9de29213db943b/pkg/api/pod/util.go#L537-L546

https://github.com/kubernetes/kubernetes/blob/1eb2acca99fefad25a29870ffc9de29213db943b/pkg/api/pod/util_test.go#L1104-L1198

**Comment by freehan** [comment]
Thank you pointing that out!

**Comment by liggitt** [comment]
/retest

mechanics look good, will defer to @thockin on API promotion

**Review by thockin — COMMENTED** [review]
Before I can approve this, I think we need to have the discussion about polarity of conditions.  We started in email, but we need to resolve it.

TL;DR We document that conditions should be "bad", so that True means something unusual.  This uses True as a good thing.

1) We could flip this feature (and drag it through a beta 2 and other changes) to align with that (condition must be present AND False).  E.g. "LoadBalancerReady" now becomes "WaitingForLoadBalancer".

2) We could change our conventions to say that polarity is situational, but "True = bad" is preferred.

3) We could add an optional polarity flag to conditions and let implementors define them how they want.

@liggitt @smarterclayton @lavalamp - anyone have thoughts?

**Comment by freehan** [comment]
> We could flip this feature (and drag it through a beta 2 and other changes) to align with that (condition must be present AND False). E.g. "LoadBalancerReady" now becomes "WaitingForLoadBalancer".

For a boolean condition, it feels more intuitive to default it to "False" when condition is not present. 

I do not think we can just flip this feature with the existing field. We will need to add new field to conform with the convention and deprecate the old field and allow users to migrate in between. 

> We could change our conventions to say that polarity is situational, but "True = bad" is preferred.

I am not sure why we added the polarity recommendation in API convention in the first place. But my understanding is that it falls under this assumption: Because an object would not know what are the possible conditions out there. If additional conditions present, it should indicate problems and the object should be considered in "bad" state. If no conditions present, the object is considered "good". If we do the reverse, it would be hard/racy to determine if an object is in "good" state because other conditions may still pop up. 

Do I understand it correctly?

However, for PodReadinessGate feature, we defined ReadinessGates in PodSpec and disallow updates, so the corresponding conditions are expected to be presented. Hence, I feel like the assumption above does not hold true.

**Comment by liggitt** [comment]
> For a boolean condition, it feels more intuitive to default it to "False" when condition is not present.

I agree. For things that must be present to indicate readiness, I map "all systems go" to true values much more intuitively. "Ready=true" is immediately understandable. "NotReady=false" makes me go into a slow-motion double-negative headspace.

Keeping `readinessGates` the same polarity as the pod `Ready` condition also seems more coherent to me.

> We could change our conventions to say that polarity is situational, but "True = bad" is preferred.

I do think it is situational. One rationalization might be to consider whether absence of a condition is considered good or bad, and make that align with an explicit false value:
* for readinessGates, absence is bad, so absence/false=bad, true=good
* for node conditions like disk pressure, absence is assumed to be ok, so absence/false=good, true=bad

**Comment by soggiest** [comment]
Hello! Code freeze for 1.14 is coming up in 2 days. Will this be merged in the next week?

@liggitt @thockin

**Comment by thockin** [comment]
The guidance on polarity was added in part so that people could reason
about unknown conditions - you don't have a condition, so you must be OK.
We even cited Ready as a counter-example.  This ability to reason about
things consistently is pretty powerful.  I don't need to know what I don't
know.  No news is good news.

I agree that conditions which don't exist being "False" makes most sense.
That aligns well with "conditions are bad things" ethos.  This feature
violates the ethos.  Now I need to know what I don't know (which is why
readinessGates the field exists).  Hypothetically, we could have the Pod
defaulting add each entry in `readinessGates` to Conditions  as "True".
E.g. you would see "WaitingForLoadBalancer" = "True" at pod creation time.
The LB controller would change it to false (or even delete it).

If we're OK to let this proceed, and I am inclined to do so, we need to
change the API conventions.  We should consider a "hint" field for polarity.

**Comment by thockin** [comment]
To expedite this, I spoke with Jordan some more.  The conclusion is that we both agree that the UX of conditions is important and in this case it makes more sense to keep the polarity as-is.  If I had a time machine, I might have argued harder to design this to be polarity-consistent with guidance.  Alas, my time machine is in the shop until last week.

/lgtm
/approve
